File: backend/src/utils/query.ts
export function queryValues(value: unknown): string[] {
  const raw = Array.isArray(value) ? value : value == null ? [] : [value];
  const unique: string[] = [];
  for (const item of raw) {
    if (typeof item !== "string") continue;
    for (const piece of item.split(",")) {
      const trimmed = piece.trim();
      if (trimmed && !unique.some((existing) => existing.toLowerCase() === trimmed.toLowerCase())) {
        unique.push(trimmed);
      }
    }
  }
  return unique;
}

function firstString(value: unknown): string | undefined {
  if (Array.isArray(value)) return firstString(value[0]);
  return typeof value === "string" ? value : undefined;
}

export function queryParam(
  req: { query: Record<string, unknown>; url?: string; originalUrl?: string },
  key: string,
): string | undefined {
  const raw = req.originalUrl ?? req.url ?? "";
  const q = raw.includes("?") ? raw.slice(raw.indexOf("?") + 1) : "";
  if (q) {
    const fromUrl = new URLSearchParams(q).get(key);
    if (fromUrl) return fromUrl;
  }
  return firstString(req.query[key]);
}

File: backend-py/app/query.py
from __future__ import annotations


def query_values(value: str | list[str] | None) -> list[str]:
    if value is None:
        return []
    raw = value if isinstance(value, list) else [value]
    unique: list[str] = []
    for item in raw:
        if not isinstance(item, str):
            continue
        for piece in item.split(","):
            trimmed = piece.strip()
            if trimmed and not any(existing.lower() == trimmed.lower() for existing in unique):
                unique.append(trimmed)
    return unique


def parse_positive_int(value: str | None, fallback: int, maximum: int = 2**53 - 1) -> int:
    if value is None:
        return fallback
    try:
        n = int(value)
    except (TypeError, ValueError):
        return fallback
    if n < 1:
        return fallback
    return min(maximum, n)
